From d089845af2b0b3ea0d441e79799dbc4d20b9b1b4 Mon Sep 17 00:00:00 2001
From: "Thulasimani,Sivakumar" <sivakumar.thulasimani@intel.com>
Date: Tue, 15 Sep 2015 20:05:05 +0530
Subject: [PATCH] SQUASHME [VPG]: drm/i915: Dis/Enable Phy in upfront
 linktraining

This patch disables and enables Phy layer which is required for next
enable request to succeed. Upfront link training will retry link
training at lower lane count and link rate values if it failed with
existing values, this requires us to disable and enable the phy layer,
which is done in this patch.

SQUASHME! - This patch should be squashed into the following existing patch:
Author:    Durgadoss R <durgadoss.r@intel.com>
Date:      Fri, 22 May 2015 14:30:07 +0530
Commit:    45cb921339e44f6c2c8ea7bf5b21f3d384769296
Change-Id: I2ae866b32a0998abcf32826efa2504e203dc8ffa
Subject:   FOR_UPSTREAM [VPG]: drm/i915: Enable Upfront link training
           for type-C DP support

Change-Id: I0a681c7f32bf7103aa488a7eab6befad885ad03c
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-5971
Signed-off-by: Sivakumar Thulasimani <sivakumar.thulasimani@intel.com>
---
 drivers/gpu/drm/i915/intel_display.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index 5d818a1..a469f64 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -14997,9 +14997,11 @@ bool chv_upfront_link_train(struct drm_device *dev,
 	struct drm_i915_private *dev_priv = dev->dev_private;
 	struct intel_connector *connector = intel_dp->attached_connector;
 	struct intel_encoder *encoder = connector->encoder;
+	enum intel_display_power_domain power_domain;
 	bool found = false;
 	bool valid_crtc = false;
 	uint8_t tmp_lane_count, tmp_link_bw;
+	power_domain = intel_display_port_power_domain(encoder);
 
 	if (!connector || !encoder) {
 		DRM_DEBUG_KMS("dp connector/encoder is NULL\n");
@@ -15089,6 +15091,13 @@ start_link_train:
 		DRM_DEBUG_KMS("upfront link training failed. lanes:%d bw:%d\n",
 				intel_dp->lane_count, intel_dp->link_bw);
 
+		/*
+		 * Phy layer needs to be turned off and on for
+		 * new set enabling to work
+		 */
+		intel_display_power_put(dev_priv, power_domain);
+		intel_display_power_get(dev_priv, power_domain);
+
 		/* Go down to the next level and retry link training */
 		if (intel_dp->lane_count == 4) {
 			intel_dp->lane_count = 2;
-- 
1.9.1

